packages/caravan-health/src/privacy.ts

```diff
@@ -0,0 +1,285 @@
+}
+
+/*
+The methodology for calculating a privacy score (p_score) for Bitcoin transactions based 
+on the number of inputs and outputs is the primary point to define wallet health for privacy. 
+The score is further influenced by several factors such as address reuse, 
+address types and UTXO set fingerprints etc.
+
+More on the algorithms for scoring privacy: <TODO : Add Link to Blog Post>
+*/
+
+// A normalizing quantity that increases the score by a certain factor in cases of self-payment.
+// More about deniability : https://www.truthcoin.info/blog/deniability/
+const DENIABILITY_FACTOR = 1.5;
+
+/*
+The p_score is calculated by evaluating the likelihood of self-payments, the involvement of 
+change outputs and the type of transaction based on number of inputs and outputs.
+
+We have 5 categories of transaction type
+- Sweep Spend
+- Simple Spend
+- UTXO Fragmentation
+- Consolidation
+- CoinJoin
+*/
+enum SpendType {
+  SweepSpend = "SweepSpend",
+  SimpleSpend = "SimpleSpend",
+  UTXOFragmentation = "UTXOFragmentation",
+  Consolidation = "Consolidation",
+  MixingOrCoinJoin = "MixingOrCoinJoin",
+}
+
+function SpendTypeScores(
+  spendType: SpendType,
+  numberOfInputs: number,
+  numberOfOutputs: number,
+): number {
+  switch (spendType) {
+    case SpendType.SweepSpend:
+      return 1 / 2;
+    case SpendType.SimpleSpend:
+      return 4 / 9;
+    case SpendType.UTXOFragmentation:
+      return 2 / 3 - 1 / numberOfOutputs;
+    case SpendType.Consolidation:
+      return 1 / numberOfInputs;
+    case SpendType.MixingOrCoinJoin:
+      let x = Math.pow(numberOfOutputs, 2) / numberOfInputs;
+      return (0.75 * x) / (1 + x);
+  }
+}
+
+function determineSpendType(inputs: number, outputs: number): SpendType {
+  if (inputs === 1) {
+    if (outputs === 1) return SpendType.SweepSpend;
+    if (outputs === 2) return SpendType.SimpleSpend;
+    return SpendType.UTXOFragmentation;
+  } else {
+    if (outputs === 1) return SpendType.Consolidation;
+    return SpendType.MixingOrCoinJoin;
+  }
+}
+
+export function privscyScoreByTxTopology(
+  transaction: Transaction,
+  client: BlockchainClient,
+): number {
+  const numberOfInputs: number = transaction.vin.length;
+  const numberOfOutputs: number = transaction.vout.length;
+
+  const spendType: SpendType = determineSpendType(
+    numberOfInputs,
+    numberOfOutputs,
+  );
+  const score: number = SpendTypeScores(
+    spendType,
+    numberOfInputs,
+    numberOfOutputs,
+  );
+  if (
+    isSelfPayment(transaction, client) &&
+    spendType !== SpendType.Consolidation
+  ) {
+    return score * DENIABILITY_FACTOR;
+  }
+  return score;
+}
+
+/*
+Determine whether a Bitcoin transaction is a self-payment, meaning the sender and recipient
+are the same entity. To check this, we make an RPC call to the watcher wallet asking for the
+amount that a given address holds. If the call returns a number then it is part of wallet
+otherwise it is not a self payment.
+*/
+function isSelfPayment(
```

Comment:
this function is going to run into issues. Since you have an async call in the `forEach` but it is non-blocking so the this will always return true. 

You should instead write out a for...of loop and make this function async. Something like this:

```
async function isSelfPayment(
  transaction: Transaction,
  client: BlockchainClient,
): Promise<boolean> {
  for (const op of transaction.vout) {
    if ((await client.getAddressStatus(op.scriptPubkeyAddress)) === undefined) {
      return false;
    }
  }
  return true;
}
```